Add comprehensive documentation to README and improve PHP docblocks (#7)
jahidulpabelislam/crud-api: README.md

[![Total Downloads](https://poser.pugx.org/jpi/crud-api/downloads)](https://packagist.org/packages/jpi/crud-api)
[![Latest Unstable Version](https://poser.pugx.org/jpi/crud-api/v/unstable)](https://packagist.org/packages/jpi/crud-api)
[![Licence](https://poser.pugx.org/jpi/crud-api/license)](https://packagist.org/packages/jpi/crud-api)
![GitHub last commit (branch)](https://img.shields.io/github/last-commit/jahidulpabelislam/crud-api/1.x.svg?label=last%20activity)

A lightweight PHP framework for building RESTful CRUD APIs with built-in support for pagination, search, filtering, and authentication.

## Features

- **RESTful CRUD Operations**: Out-of-the-box support for List, Create, Read, Update and Delete operations
- **Pagination**: Built-in pagination support with configurable page size
- **Search & Filtering**: Flexible search and filter functionality for listing endpoints
- **Authentication**: Built-in authentication checks for protected endpoints
- **JSON Responses**: Standardised JSON response format with HATEOAS links
- **Validation**: Comprehensive data validation with detailed error messages
- **Extensible**: Easy to extend for custom business logic

> ## Requirements

Reviewer: ```suggestion
## Dependencies
```

    ];

    public function getAPIURL(): \JPI\Utils\URL {
        return new \JPI\Utils\URL("https://api.example.com/projects/{$this->getId()}");
    }
}
```

### 2. Create a Controller

**AbstractController** is the base controller class providing standard CRUD endpoints with authentication support.

**Standard Actions:**
- `index()`: GET - List all entities (with pagination, search, and filters)
- `create()`: POST - Create a new entity
- `read($id)`: GET - Retrieve a specific entity
- `update($id)`: PUT - Update a specific entity
- `delete($id)`: DELETE - Delete a specific entity

The **Responder** trait provides standardised response methods for controllers including item responses with proper status codes, collection responses with pagination metadata, and error responses.

```php
<?php

class ProjectController extends \JPI\CRUD\API\AbstractController {
    
    protected string $entityClass = Project::class;
    
    // Define which actions are publicly accessible (no authentication required)
    protected array $publicActions = ["index", "read"];
}
```

### 3. Set Up Routes

**Router** is an extended HTTP router with built-in 404 and 405 error handlers.

> You can set up CRUD routes manually or use the convenient `addCRUDRoutes()` method:

Reviewer: @copilot Explain what routes this will add and what controller methods it will point to.

Reviewer: Thanks. @copilot Can you also add note about the $name argument.

```php
<?php

$request = \JPI\HTTP\Request::createFromGlobals();
$router = new \JPI\CRUD\API\Router($request);

// Option 1: Use addCRUDRoutes() helper (recommended)
$router->addCRUDRoutes("/projects", ProjectController::class);

// Option 2: Define routes manually

> $router->group("/projects", function($router) {

Reviewer: @copilot There isn't a group method, nor the methods below. It should be `addRoute` and the method passed as 2nd argument.

    $router->get("/", [ProjectController::class, "index"]);
    $router->post("/", [ProjectController::class, "create"]);
    $router->get("/{id}/", [ProjectController::class, "read"]);
    $router->put("/{id}/", [ProjectController::class, "update"]);
    $router->delete("/{id}/", [ProjectController::class, "delete"]);
});

> $response = $router->dispatch($request);

Reviewer: ```suggestion
$response = $router->handle();
```

@copilot

    }
}
```

### 2. Create a Controller

**AbstractController** is the base controller class providing standard CRUD endpoints with authentication support.

**Standard Actions:**
- `index()`: GET - List all entities (with pagination, search, and filters)
- `create()`: POST - Create a new entity
- `read($id)`: GET - Retrieve a specific entity
- `update($id)`: PUT - Update a specific entity
- `delete($id)`: DELETE - Delete a specific entity

The **Responder** trait provides standardised response methods for controllers including item responses with proper status codes, collection responses with pagination metadata, and error responses.

```php
<?php

class ProjectController extends \JPI\CRUD\API\AbstractController {
    
    protected string $entityClass = Project::class;
    
    // Define which actions are publicly accessible (no authentication required)
    protected array $publicActions = ["index", "read"];
}
```

### 3. Set Up Routes

**Router** is an extended HTTP router with built-in 404 and 405 error handlers.

You can set up CRUD routes manually or use the convenient `addCRUDRoutes()` method:

```php
<?php

$request = \JPI\HTTP\Request::createFromGlobals();

> $router = new \JPI\CRUD\API\Router($request);

Reviewer: @copilot Need to initialise `App` too. And `handle` is called on the `$app` rather than `$router`.

You can set up CRUD routes manually or use the convenient `addCRUDRoutes()` method:

```php
<?php

$request = \JPI\HTTP\Request::createFromGlobals();
$router = new \JPI\CRUD\API\Router($request);
$app = new \JPI\CRUD\API\App($router);

// Option 1: Use addCRUDRoutes() helper (recommended)
// This creates 5 routes:
//   GET    /projects/     -> ProjectController::index
//   POST   /projects/     -> ProjectController::create
//   GET    /projects/{id}/ -> ProjectController::read
//   PUT    /projects/{id}/ -> ProjectController::update
//   DELETE /projects/{id}/ -> ProjectController::delete
// Optional third parameter: route name for the read action
$app->addCRUDRoutes("/projects", ProjectController::class);
// Or with a named route: $app->addCRUDRoutes("/projects", ProjectController::class, "project");

// Option 2: Define routes manually
$router->addRoute("/projects/", "GET", ProjectController::class . "::index");
$router->addRoute("/projects/", "POST", ProjectController::class . "::create");
$router->addRoute("/projects/{id}/", "GET", ProjectController::class . "::read");
$router->addRoute("/projects/{id}/", "PUT", ProjectController::class . "::update");

> $router->addRoute("/projects/{id}/", "DELETE", ProjectController::class . "::delete");

Reviewer: ```suggestion
$app->addRoute("/projects/", "GET", ProjectController::class . "::index");
$app->addRoute("/projects/", "POST", ProjectController::class . "::create");
$app->addRoute("/projects/{id}/", "GET", ProjectController::class . "::read");
$app->addRoute("/projects/{id}/", "PUT", ProjectController::class . "::update");
$app->addRoute("/projects/{id}/", "DELETE", ProjectController::class . "::delete");
```

```php
<?php

class ProjectController extends \JPI\CRUD\API\AbstractController {
    
    protected string $entityClass = Project::class;
    
    // Define which actions are publicly accessible (no authentication required)
    protected array $publicActions = ["index", "read"];
}
```

### 3. Set Up Routes

```php
<?php

$request = \JPI\HTTP\Request::createFromGlobals();
$router = new \JPI\CRUD\API\Router($request);

$router->group("/projects", function($router) {
    $router->get("/", [ProjectController::class, "index"]);
    $router->post("/", [ProjectController::class, "create"]);
    $router->get("/{id}/", [ProjectController::class, "read"]);
    $router->put("/{id}/", [ProjectController::class, "update"]);
    $router->delete("/{id}/", [ProjectController::class, "delete"]);
});

$response = $router->dispatch($request);
$response->send();
```

### 4. Customise CRUD Service (Optional)

```php
<?php

class ProjectService extends \JPI\CRUD\API\CrudService {
    
    protected bool $paginated = true;
    protected int $perPage = 20;
    
    protected static array $requiredColumns = ["name"];
}
```

Then reference it in your entity's `crudService` property:

> ```php

Reviewer: @copilot No need for this super basic example.

A lightweight PHP framework for building RESTful CRUD APIs with built-in support for pagination, search, filtering, and authentication.

## Features

- **RESTful CRUD Operations**: Out-of-the-box support for Create, Read, Update, and Delete operations
- **Pagination**: Built-in pagination support with configurable page size
- **Search & Filtering**: Flexible search and filter functionality using traits
- **Authentication**: Built-in authentication checks for protected endpoints
- **Type Safety**: Full PHP 8+ type declarations for improved code quality
- **JSON Responses**: Standardised JSON response format with HATEOAS links
- **Validation**: Comprehensive data validation with detailed error messages
- **Extensible**: Easy to extend with custom business logic

## Requirements

- PHP 8.0+
- Composer
- [jpi/http](https://packagist.org/packages/jpi/http) v1
- [jpi/orm](https://packagist.org/packages/jpi/orm) v2
- [jpi/utils](https://packagist.org/packages/jpi/utils) v1

## Installation

Install via [Composer](https://getcomposer.org/):

```bash
composer require jpi/crud
```

## Core Components

> ### AbstractEntity

Reviewer: @copilot Feels like the Components could be merged in with related bit in `Basic Usage`, what do you think?

Reviewer: @copilot Yes please.

    "_total_pages": 3,
    "_links": {
        "self": "https://api.example.com/projects?page=1",
        "next_page": "https://api.example.com/projects?page=2"
    }
}
```

### Error Response

```json
{
    "message": "The necessary data was not provided and/or invalid.",
    "errors": {
        "name": "`name` is required.",
        "email": "Must be a valid email address."
    }
}
```

## Advanced Features

### Search

Enable search on your entities by implementing `SearchableInterface` and using the `Searchable` trait, then adding `searchableColumns` property to your Entity.

Then use the search query parameter:
```
GET /projects?search=web+development
```

### Filtering

Enable filtering by implementing `FilterableInterface` and using the `Filterable` trait, then adding `filterableColumns` property to your Entity.

Then use the filters query parameter:
```
GET /projects?filters[status]=active&filters[category]=web
```

### Authentication

The framework checks for an `is_authenticated` request attribute. Integrate with your authentication system by setting this attribute.

Endpoints not listed in `$publicActions` will return a 401 response if the user is not authenticated.

> 

Reviewer: ```suggestion
```

    "_links": {
        "self": "https://api.example.com/projects?page=1",
        "next_page": "https://api.example.com/projects?page=2"
    }
}
```

### Error Response

```json
{
    "message": "The necessary data was not provided and/or invalid.",
    "errors": {
        "name": "`name` is required.",
        "email": "Must be a valid email address."
    }
}
```

## Advanced Features

### Search

Enable search on your entities by implementing `SearchableInterface` and using the `Searchable` trait, then adding `searchableColumns` property to your Entity.

Then use the search query parameter:
```
GET /projects?search=web+development
```

### Filtering

Enable filtering by implementing `FilterableInterface` and using the `Filterable` trait, then adding `filterableColumns` property to your Entity.

Then use the filters query parameter:
```
GET /projects?filters[status]=active&filters[category]=web
```

### Authentication

The framework checks for an `is_authenticated` request attribute. Integrate with your authentication system by setting this attribute.

Endpoints not listed in `$publicActions` will return a 401 response if the user is not authenticated.

## Examples

> See a real-world implementation in the [Portfolio API](https://github.com/jahidulpabelislam/portfolio-api) project.

Reviewer: ```suggestion
```

**Standard Actions:**
- `index()`: GET - List all entities (with pagination, search, and filters)
- `create()`: POST - Create a new entity
- `read($id)`: GET - Retrieve a specific entity
- `update($id)`: PUT - Update a specific entity
- `delete($id)`: DELETE - Delete a specific entity

### CrudService

Service layer handling CRUD operations and data validation.

**Key Features:**
- Automatic request data validation
- Support for required fields
- Search and filter integration
- Pagination support

### Router

Extended HTTP router with built-in 404 and 405 error handlers.

### Entity Traits

**Searchable**: Adds search functionality to entities
- Define searchable columns via `$searchableColumns` property
- Multi-word search support

**Filterable**: Adds filtering functionality to entities
- Define filterable columns via `$filterableColumns` property
- Equality-based filtering

**Responder**: Provides standardised response methods for controllers
- Item responses with proper status codes
- Collection responses with pagination metadata
- Error responses (not found, validation errors)

## Basic Usage

### 1. Create an Entity

```php
<?php

> namespace App;

Reviewer: @copilot Don't have a namespace in the examples.

    
    // Define which actions are publicly accessible (no authentication required)
    protected array $publicActions = ["index", "read"];
}
```

### 3. Set Up Routes

```php
<?php

$request = \JPI\HTTP\Request::createFromGlobals();
$router = new \JPI\CRUD\API\Router($request);

$router->group("/projects", function($router) {
    $router->get("/", [ProjectController::class, "index"]);
    $router->post("/", [ProjectController::class, "create"]);
    $router->get("/{id}/", [ProjectController::class, "read"]);
    $router->put("/{id}/", [ProjectController::class, "update"]);
    $router->delete("/{id}/", [ProjectController::class, "delete"]);
});

$response = $router->dispatch($request);
$response->send();
```

### 4. Customise CRUD Service (Optional)

```php
<?php

namespace App;

class ProjectService extends \JPI\CRUD\API\CrudService {
    
    protected bool $paginated = true;
    protected int $perPage = 20;
    
    protected static array $requiredColumns = ["name"];
}
```

Then reference it in your entity:

```php
class Project extends \JPI\CRUD\API\AbstractEntity {
    protected static string $crudService = ProjectService::class;
    // ...
}

> ```

Reviewer: ```suggestion
Then reference it in your entity's `crudService` property:
```

@copilot

```json
{
    "data": [
        {
            "id": 1,
            "name": "Project One",
            "_links": {
                "self": "https://api.example.com/projects/1"
            }
        },
        {
            "id": 2,
            "name": "Project Two",
            "_links": {
                "self": "https://api.example.com/projects/2"
            }
        }
    ],
    "_total_count": 25,
    "_total_pages": 3,
    "_links": {
        "self": "https://api.example.com/projects?page=1",
        "next_page": "https://api.example.com/projects?page=2"
    }
}
```

### Error Response

```json
{
    "message": "The necessary data was not provided and/or invalid.",
    "errors": {
        "name": "`name` is required.",
        "email": "Must be a valid email address."
    }
}
```

## Advanced Features

### Search

Enable search on your entities by implementing `SearchableInterface` and using the `Searchable` trait:

```php
class Project extends \JPI\CRUD\API\AbstractEntity implements \JPI\CRUD\API\Entity\SearchableInterface {
    use \JPI\CRUD\API\Entity\Searchable;
    
    protected static array $searchableColumns = ["name", "description"];
}

> ```

Reviewer: ```suggestion
Enable search on your entities by implementing `SearchableInterface` and using the `Searchable` trait, the adding `searchableColumns` property to your Entity.
```

Reviewer: @copilot

Then use the search query parameter:
```
GET /projects?search=web+development
```

### Filtering

Enable filtering by implementing `FilterableInterface` and using the `Filterable` trait:

```php
class Project extends \JPI\CRUD\API\AbstractEntity implements \JPI\CRUD\API\Entity\FilterableInterface {
    use \JPI\CRUD\API\Entity\Filterable;
    
    protected static array $filterableColumns = ["status", "category"];
}

> ```

Reviewer: ```suggestion
Enable filtering by implementing `FilterableInterface` and using the `Filterable` trait, then adding `filterableColumns` property to your Entity.
```

@copilot

Then use the filters query parameter:
```
GET /projects?filters[status]=active&filters[category]=web
```

### Authentication

The framework checks for an `is_authenticated` request attribute. Integrate with your authentication system by setting this attribute:

> ```php

Reviewer: @copilot Remove example

[![Licence](https://poser.pugx.org/jpi/crud/license)](https://packagist.org/packages/jpi/crud)
![GitHub last commit (branch)](https://img.shields.io/github/last-commit/jahidulpabelislam/crud-api/1.x.svg?label=last%20activity)

A lightweight PHP framework for building RESTful CRUD APIs with built-in support for pagination, search, filtering, and authentication.

## Features

- **RESTful CRUD Operations**: Out-of-the-box support for Create, Read, Update, and Delete operations
- **Pagination**: Built-in pagination support with configurable page size
- **Search & Filtering**: Flexible search and filter functionality using traits
- **Authentication**: Built-in authentication checks for protected endpoints
- **Type Safety**: Full PHP 8+ type declarations for improved code quality
- **JSON Responses**: Standardised JSON response format with HATEOAS links
- **Validation**: Comprehensive data validation with detailed error messages
- **Extensible**: Easy to extend with custom business logic

## Requirements

- PHP 8.0+
- Composer

> - [jpi/http](https://github.com/jahidulpabelislam/http) - HTTP request/response handling

Reviewer: Use links to packagist where possible and add version requirments.

**Standard Actions:**
- `index()`: GET - List all entities (with pagination, search, and filters)
- `create()`: POST - Create a new entity
- `read($id)`: GET - Retrieve a specific entity
- `update($id)`: PUT - Update a specific entity
- `delete($id)`: DELETE - Delete a specific entity

### CrudService

Service layer handling CRUD operations and data validation.

**Key Features:**
- Automatic request data validation
- Support for required fields
- Search and filter integration
- Pagination support

### Router

Extended HTTP router with built-in 404 and 405 error handlers.

### Entity Traits

**Searchable**: Adds search functionality to entities
- Define searchable columns via `$searchableColumns` property
- Multi-word search support

**Filterable**: Adds filtering functionality to entities
- Define filterable columns via `$filterableColumns` property
- Equality-based filtering

**Responder**: Provides standardised response methods for controllers
- Item responses with proper status codes
- Collection responses with pagination metadata
- Error responses (not found, validation errors)

## Basic Usage

### 1. Create an Entity

```php
<?php

> namespace App;

Reviewer: Don't import classes in code examples, use full namespace.

Example `class Project extends \JPI\CRUD\API\AbstractEntity..`

    use Searchable;
    
    protected static array $searchableColumns = ["name", "description"];
}
```

Then use the search query parameter:
```
GET /projects?search=web+development
```

### Filtering

Enable filtering by implementing `FilterableInterface` and using the `Filterable` trait:

```php
use JPI\CRUD\API\Entity\FilterableInterface;
use JPI\CRUD\API\Entity\Filterable;

class Project extends AbstractEntity implements FilterableInterface {
    use Filterable;
    
    protected static array $filterableColumns = ["status", "category"];
}
```

Then use the filters query parameter:
```
GET /projects?filters[status]=active&filters[category]=web
```

### Authentication

The framework checks for an `is_authenticated` request attribute. Integrate with your authentication system by setting this attribute:

```php
$request = $request->withAttribute("is_authenticated", $isUserLoggedIn);
```

Endpoints not listed in `$publicActions` will return a 401 response if the user is not authenticated.

## Examples

See a real-world implementation in the [Portfolio API](https://github.com/jahidulpabelislam/portfolio-api) project.

## Support

> If you found this library useful, please consider:

Reviewer: Reverted changes to this section pls

## Authors

> - [Jahidul Pabel Islam](https://jahidulpabelislam.com/) — [me@jahidulpabelislam.com](mailto:me@jahidulpabelislam.com)

Reviewer: ```suggestion
- [Jahidul Pabel Islam](https://jahidulpabelislam.com/) [<me@jahidulpabelislam.com>](mailto:me@jahidulpabelislam.com)
```

## Licence

> This module is licensed under the GNU General Public Licence v3.0 — see the [licence](LICENSE.md) file for details.

Reviewer: ```suggestion
This module is licensed under the GNU General Public Licence — see the [licence](LICENSE.md) file for details.
```

## Authors

- [Jahidul Pabel Islam](https://jahidulpabelislam.com/) — [me@jahidulpabelislam.com](mailto:me@jahidulpabelislam.com)

## Licence

> This project is licenced under the GNU General Public Licence v3.0 — see the [LICENSE](LICENSE.md) file for details.

Reviewer: ```suggestion
This module is licensed under the GNU General Public Licence v3.0 — see the [licence](LICENSE.md) file for details.
```
